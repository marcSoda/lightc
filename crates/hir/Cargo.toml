[package]
name = "hir"
version = "0.0.0"
edition = "2021"

[dependencies]
ast = { path = "../ast" }
common = { path = "../common" }

serde = { version = "1.0.136", features = ["derive"] }

[dev-dependencies]
lexer = { path = "../lexer" }
parser = { path = "../parser" }

insta = { version = "1.12.0", features = ["glob"] }
