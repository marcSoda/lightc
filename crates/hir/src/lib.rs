use ast::{Ast, AstNode, AstVisitor, Visitable};
use common::Operator;
pub use node::{HirNode, NodeKind};
use symbol_table::{Symbol, SymbolTable, Symbolic};

mod node;
#[cfg(test)]
mod tests;

// Performs the following:
// - desugars x += 1 to x = x + 1
// - cooks function names in the AST and symbol table
// - tracks scope (needed?)

type HirResult = Result<HirNode, String>;

pub struct Hir<'a> {
    symbol_table: &'a mut SymbolTable<Symbol>,
}

impl<'a> AstVisitor for Hir<'a> {
    type AstNode = AstNode;
    type Result = Result<HirNode, String>;

    fn visit_for(&mut self, s: ast::For<Self::AstNode>) -> Self::Result {
        self.lower_for(s)
    }

    fn visit_let(&mut self, s: ast::Let<Self::AstNode>) -> Self::Result {
        self.lower_let(s)
    }

    fn visit_fn(&mut self, s: ast::Fn<Self::AstNode>) -> Self::Result {
        self.lower_func(s)
    }

    fn visit_struct(&mut self, s: ast::Struct<Self::AstNode>) -> Self::Result {
        self.lower_struct(s)
    }

    fn visit_lit(&mut self, e: ast::Lit<Self::AstNode>) -> Self::Result {
        self.lower_lit(e)
    }

    fn visit_binop(&mut self, e: ast::BinOp<Self::AstNode>) -> Self::Result {
        self.lower_binop(e)
    }

    fn visit_unop(&mut self, e: ast::UnOp<Self::AstNode>) -> Self::Result {
        self.lower_unop(e)
    }

    fn visit_ident(&mut self, e: ast::Ident) -> Self::Result {
        self.lower_ident(e)
    }

    fn visit_call(&mut self, e: ast::Call<Self::AstNode>) -> Self::Result {
        self.lower_call(e)
    }

    fn visit_cond(&mut self, e: ast::Cond<Self::AstNode>) -> Self::Result {
        self.lower_cond(e)
    }

    fn visit_block(&mut self, e: ast::Block<Self::AstNode>) -> Self::Result {
        self.lower_block(e)
    }

    fn visit_index(&mut self, e: ast::Index<Self::AstNode>) -> Self::Result {
        self.lower_index(e)
    }
}

impl<'a> Hir<'a> {
    pub fn new(symbol_table: &'a mut SymbolTable<Symbol>) -> Self {
        Hir { symbol_table }
    }

    pub fn walk(mut self, ast: Ast<AstNode>) -> Result<Ast<HirNode>, String> {
        let mut hir = Ast::new();
        for node in ast.into_nodes() {
            let lowered_node = node.accept(&mut self)?;
            hir.add(lowered_node)
        }
        Ok(hir)
    }

    // XXX replace with accept()
    fn lower_node(&mut self, node: AstNode) -> Result<HirNode, String> {
        use ast::NodeKind::*;

        Ok(match node.kind {
            For(s) => self.lower_for(s)?,
            Let(s) => self.lower_let(s)?,
            Fn(s) => self.lower_func(s)?,
            Struct(s) => self.lower_struct(s)?,
            Lit(e) => self.lower_lit(e)?,
            Ident(e) => self.lower_ident(e)?,
            BinOp(e) => self.lower_binop(e)?,
            UnOp(e) => self.lower_unop(e)?,
            Call(e) => self.lower_call(e)?,
            Cond(e) => self.lower_cond(e)?,
            Block(e) => self.lower_block(e)?,
            Index(e) => self.lower_index(e)?,
        })
    }

    fn lower_for(&mut self, stmt: ast::For<AstNode>) -> HirResult {
        // Insert start var
        self.symbol_table.enter_scope();
        self.symbol_table.insert(Symbol::new_var(&stmt.start_name, &stmt.start_antn));

        let start_expr = self.lower_var_init(&stmt.start_name, stmt.start_expr.as_deref())?;
        let cond_expr = self.lower_node(*stmt.cond_expr)?;
        let step_expr = self.lower_node(*stmt.step_expr)?;
        let body = self.lower_node(*stmt.body)?;

        self.symbol_table.leave_scope();

        Ok(HirNode::new_for(stmt.start_name, stmt.start_antn, Some(start_expr), cond_expr, step_expr, body))
    }

    fn lower_let(&mut self, stmt: ast::Let<AstNode>) -> HirResult {
        self.symbol_table.insert(Symbol::new_var(&stmt.name, &stmt.antn));
        let init_node = self.lower_var_init(&stmt.name, stmt.init.as_deref())?;
        Ok(HirNode::new_let(stmt.name, stmt.antn, Some(init_node)))
    }

    fn lower_func(&mut self, mut stmt: ast::Fn<AstNode>) -> HirResult {
        // Insert a duplicate of the symbol. The new one will have the lowered
        // name. Update the AST as well. Skip for externs.
        let sym =
            self.symbol_table.get(stmt.proto.name()).cloned().unwrap_or_else(|| {
                unreachable!("missing symbol in `lower_func()` for `{}`", stmt.proto.name())
            });

        if !sym.is_extern() {
            stmt.proto.set_name(sym.name().to_owned());
            self.symbol_table.insert(sym);
        }

        // This creates an interstitial scope for the arguments in the function definition
        // because lower_block() will also create a new scope. Shouldn't be a practical
        // issue.
        self.symbol_table.enter_scope();

        for arg in stmt.proto.args() {
            self.symbol_table.insert(Symbol::new_var(&arg.0, &arg.1));
        }

        let body_node = stmt.body.map(|e| self.lower_node(*e));

        self.symbol_table.leave_scope();

        Ok(HirNode::new_fn(*stmt.proto, body_node.transpose()?))
    }

    fn lower_struct(&mut self, _stmt: ast::Struct<AstNode>) -> HirResult {
        todo!()
    }

    fn lower_lit(&mut self, expr: ast::Lit<AstNode>) -> HirResult {
        use ast::Literal::*;

        // Rewrapping primitives is annoying. Remove for pimitives if we dump AstNode -> HirNode
        let lit = match expr.value {
            Int8(l) => ast::Literal::Int8(l),
            Int16(l) => ast::Literal::Int16(l),
            Int32(l) => ast::Literal::Int32(l),
            Int64(l) => ast::Literal::Int64(l),
            UInt8(l) => ast::Literal::UInt8(l),
            UInt16(l) => ast::Literal::UInt16(l),
            UInt32(l) => ast::Literal::UInt32(l),
            UInt64(l) => ast::Literal::UInt64(l),
            Float(l) => ast::Literal::Float(l),
            Double(l) => ast::Literal::Double(l),
            Bool(l) => ast::Literal::Bool(l),
            Char(l) => ast::Literal::Char(l),
            Array { .. } => self.lower_lit_array(expr.value)?,
        };
        Ok(HirNode::new_lit(lit, expr.ty))
    }

    fn lower_lit_array(&mut self, lit: ast::Literal<AstNode>) -> Result<ast::Literal<HirNode>, String> {
        // Extract the elements vec and the type of the array elements.
        let (elements, ty) = match lit {
            ast::Literal::Array { elements, inner_ty } => (elements, inner_ty),
            _ => unreachable!("expected array literal"),
        };

        // Rewrap every element
        let mut chkd_elements = Vec::with_capacity(elements.len());
        for el in elements {
            chkd_elements.push(self.lower_node(el)?);
        }

        // Rebuild the literal and return the type
        Ok(ast::Literal::Array { elements: chkd_elements, inner_ty: ty })
    }

    fn lower_ident(&mut self, expr: ast::Ident) -> HirResult {
        Ok(HirNode::new_ident(expr.name, expr.ty))
    }

    // Lower `x += 1` to `x = x + 1`
    fn lower_binop(&mut self, expr: ast::BinOp<AstNode>) -> HirResult {
        use Operator::*;

        let lowered_lhs = self.lower_node(*expr.lhs)?;

        let top_op;
        let rhs = match expr.op {
            AddEq => {
                top_op = Assign;
                HirNode::new_binop(Add, lowered_lhs.clone(), self.lower_node(*expr.rhs)?, expr.ty.clone())
            },
            SubEq => {
                top_op = Assign;
                HirNode::new_binop(Sub, lowered_lhs.clone(), self.lower_node(*expr.rhs)?, expr.ty.clone())
            },
            MulEq => {
                top_op = Assign;
                HirNode::new_binop(Mul, lowered_lhs.clone(), self.lower_node(*expr.rhs)?, expr.ty.clone())
            },
            DivEq => {
                top_op = Assign;
                HirNode::new_binop(Div, lowered_lhs.clone(), self.lower_node(*expr.rhs)?, expr.ty.clone())
            },
            _ => {
                top_op = expr.op;
                self.lower_node(*expr.rhs)?
            },
        };

        Ok(HirNode::new_binop(top_op, lowered_lhs, rhs, expr.ty))
    }

    fn lower_unop(&mut self, expr: ast::UnOp<AstNode>) -> HirResult {
        Ok(HirNode::new_unop(expr.op, self.lower_node(*expr.rhs)?, expr.ty))
    }

    fn lower_call(&mut self, expr: ast::Call<AstNode>) -> HirResult {
        let sym = self
            .symbol_table
            .get(&expr.name)
            .unwrap_or_else(|| unreachable!("missing symbol in `lower_call()` for `{}`", expr.name));

        // Update the AST with the lowered name if it hasn't been done already and it's
        // not an extern call
        let lowered_name = match sym.name() {
            sym_name if !sym.is_extern() && sym_name != expr.name => sym_name.to_owned(),
            _ => expr.name,
        };

        let mut lowered_args = vec![];
        for node in expr.args {
            lowered_args.push(self.lower_node(node)?);
        }
        Ok(HirNode::new_call(lowered_name, lowered_args, expr.ty))
    }

    fn lower_cond(&mut self, expr: ast::Cond<AstNode>) -> HirResult {
        Ok(HirNode::new_cond(
            self.lower_node(*expr.cond_expr)?,
            self.lower_node(*expr.then_block)?,
            expr.else_block.map(|e| self.lower_node(*e)).transpose()?,
            expr.ty,
        ))
    }

    fn lower_block(&mut self, expr: ast::Block<AstNode>) -> HirResult {
        self.symbol_table.enter_scope();

        let mut lowered_list = vec![];
        for node in expr.list {
            lowered_list.push(self.lower_node(node)?);
        }

        self.symbol_table.leave_scope();

        Ok(HirNode::new_block(lowered_list, expr.ty))
    }

    fn lower_index(&mut self, expr: ast::Index<AstNode>) -> HirResult {
        Ok(HirNode::new_index(self.lower_node(*expr.binding)?, self.lower_node(*expr.idx)?, expr.ty))
    }

    // Helper for variable initializations
    fn lower_var_init(&mut self, name: &str, init: Option<&AstNode>) -> HirResult {
        if let Some(init) = init {
            self.lower_node(init.clone())
        } else {
            unreachable!("no initializer for variable: `{}`", name);
        }
    }
}
