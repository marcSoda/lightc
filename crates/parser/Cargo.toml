[package]
name = "parser"
version = "0.0.0"
edition = "2021"

[dependencies]
serde = { version = "1.0.136", features = ["derive"] }

ast = { path = "../ast" }
common = { path = "../common" }
symbol_table = { path = "../symbol_table" }

[dev-dependencies]
insta = { version = "1.12.0", features = ["glob"] }

lexer = { path = "../lexer" }
