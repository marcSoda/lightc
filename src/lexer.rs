use std::iter::Peekable;
use serde::Serialize;

use crate::token::{Symbol, Token, TokenType};

type LexResult = std::result::Result<Token, LexError>;

pub(crate) struct Lexer {
    stream: Peekable<StreamIter<char>>,
}

impl Lexer {
    pub fn new(input: &str) -> Self {
        Lexer {
            stream: StreamIter::new(input).peekable(),
        }
    }

    fn lex(&mut self) -> LexResult {
        let cur = match self.stream.next() {
            Some(cur) => cur,
            None => return Ok(Token::default()),
        };

        if cur.value.is_ascii_whitespace() {
            while let Some(c) = self.stream.peek() {
                if !c.value.is_ascii_whitespace() {
                    return self.lex();
                }
                self.stream.next();
            }
            return self.lex(); // Eat trailing newline
        }

        // Comments
        if cur == '/' && matches!(self.stream.peek(), Some(c) if *c == '/') {
            while let Some(c) = self.stream.next() {
                if c == '\n' {
                    return self.lex();
                }
            }
            return self.lex(); // Eat trailing comment
        }

        // Identifiers: keywords, function names, and variables
        if cur.value.is_ascii_alphabetic() {
            let mut identifier = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '_' {
                    identifier.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            let tt = match identifier.as_str() {
                "else" => TokenType::Else,
                "extern" => TokenType::Extern,
                //"u64" => TokenTypeVarType(TokenTypeU64),
                "fn" => TokenType::Fn,
                "for" => TokenType::For,
                "if" => TokenType::If,
                "let" => TokenType::Let,
                _ => TokenType::Ident(identifier),
            };

            return Ok(Token::from((tt, cur)));
        }

        // Numbers
        if cur.value.is_ascii_digit() {
            let mut num = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '.' {
                    num.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            return match num.parse() {
                Ok(n) => Ok(Token::from((TokenType::Int(n), cur))),
                Err(_) => Err(LexError::from((format!("Invalid number: {}", num), cur))),
            };
        }

        // Logical operators
        if let Some(next) = self.stream.peek() {
            match cur.value {
                '=' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::from((TokenType::Op(Symbol::Eq), cur)));
                }
                '&' if next == &'&' => {
                    self.stream.next();
                    return Ok(Token::from((TokenType::Op(Symbol::And), cur)));
                }
                '|' if next == &'|' => {
                    self.stream.next();
                    return Ok(Token::from((TokenType::Op(Symbol::Or), cur)));
                }
                _ => (),
            }
        }

        // Everything else
        let tt = match cur.value {
            '+' => TokenType::Op(Symbol::Plus),
            '-' => TokenType::Op(Symbol::Minus),
            '*' => TokenType::Op(Symbol::Mult),
            '/' => TokenType::Op(Symbol::Div),
            '^' => TokenType::Op(Symbol::Pow),
            '>' => TokenType::Op(Symbol::Gt),
            '<' => TokenType::Op(Symbol::Lt),
            '!' => TokenType::Op(Symbol::Not),
            '=' => TokenType::Op(Symbol::Assign),
            '}' => TokenType::CloseBrace,
            ')' => TokenType::CloseParen,
            ':' => TokenType::Colon,
            ',' => TokenType::Comma,
            '{' => TokenType::OpenBrace,
            '(' => TokenType::OpenParen,
            ';' => TokenType::Semicolon,
            c => {
                return Err(LexError::from((format!("Unknown character: {}", c), cur)));
            }
        };

        Ok(Token::from((tt, cur)))
    }
}

impl Iterator for Lexer {
    type Item = LexResult;

    fn next(&mut self) -> Option<Self::Item> {
        match self.lex() {
            Ok(Token {
                tt: TokenType::Eof, ..
            }) => None,
            x => Some(x),
        }
    }
}

// Provides additional context for each source character
#[derive(Debug, PartialEq)]
struct ContextPoint<T> {
    value: T,
    line: usize,
    column: usize,
}

impl<T> ContextPoint<T> {
    fn new(value: T, line: usize, column: usize) -> Self {
        ContextPoint {
            value,
            line: line + 1,
            column: column + 1,
        }
    }
}

impl PartialEq<char> for ContextPoint<char> {
    fn eq(&self, other: &char) -> bool {
        self.value == *other
    }
}

impl<T> From<(TokenType, ContextPoint<T>)> for Token {
    fn from((ty, ContextPoint { line, column, .. }): (TokenType, ContextPoint<T>)) -> Self {
        Token {
            tt: ty,
            line,
            column,
        }
    }
}

// Iterator for the source character stream. Supports line and column context.
struct StreamIter<T> {
    lines: Vec<Vec<T>>,
    line: usize,
    column: usize,
}

impl StreamIter<char> {
    fn new(input: &str) -> Self {
        StreamIter {
            lines: input
                .split_inclusive('\n') // can't use .lines() or we lose '\n'
                .map(|l| l.chars().collect())
                .collect(),
            line: 0,
            column: 0,
        }
    }
}

impl Iterator for StreamIter<char> {
    type Item = ContextPoint<char>;

    fn next(&mut self) -> Option<Self::Item> {
        let line = self.lines.get(self.line);
        let cc = line?
            .get(self.column)
            .map(|c| ContextPoint::new(*c, self.line, self.column))
            .or_else(|| {
                self.line += 1;
                self.column = 0;
                self.lines.get(self.line).and_then(|line| {
                    line.get(self.column)
                        .map(|c| ContextPoint::new(*c, self.line, self.column))
                })
            });
        self.column += 1;
        cc
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub(crate) struct LexError {
    message: String,
    line: usize,
    column: usize,
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Lexing error: {} at {}:{}",
            self.message, self.line, self.column
        )
    }
}

impl std::error::Error for LexError {}

impl<T> From<(String, ContextPoint<T>)> for LexError {
    fn from((msg, cp): (String, ContextPoint<T>)) -> Self {
        LexError {
            message: msg,
            line: cp.line,
            column: cp.column,
        }
    }
}

#[cfg(test)]
mod test {
    use insta::{assert_yaml_snapshot, glob, with_settings};
    use std::fs;

    use super::*;

    #[test]
    fn test_lexer() {
        with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            glob!("tests/inputs/lexer/*.input", |path| {
                let input = fs::read_to_string(path).unwrap();
                assert_yaml_snapshot!(Lexer::new(&input).collect::<Result<Vec<_>, _>>());
            });
        });
    }

    #[test]
    fn test_stream_next() {
        let input = "\
abc
def
ghi
";
        let mut stream = StreamIter::new(input);
        assert_eq!(ContextPoint::new('a', 0, 0), stream.next().unwrap());
        assert_eq!(ContextPoint::new('b', 0, 1), stream.next().unwrap());
        assert_eq!(ContextPoint::new('c', 0, 2), stream.next().unwrap());
        assert_eq!(ContextPoint::new('\n', 0, 3), stream.next().unwrap());
        assert_eq!(ContextPoint::new('d', 1, 0), stream.next().unwrap());
        assert_eq!(ContextPoint::new('e', 1, 1), stream.next().unwrap());
        assert_eq!(ContextPoint::new('f', 1, 2), stream.next().unwrap());
        assert_eq!(ContextPoint::new('\n', 1, 3), stream.next().unwrap());
        assert_eq!(ContextPoint::new('g', 2, 0), stream.next().unwrap());
        assert_eq!(ContextPoint::new('h', 2, 1), stream.next().unwrap());
        assert_eq!(ContextPoint::new('i', 2, 2), stream.next().unwrap());
        assert_eq!(ContextPoint::new('\n', 2, 3), stream.next().unwrap());
        assert_eq!(None, stream.next());
        assert_eq!(None, stream.next());
    }

    #[test]
    fn test_lex() {
        use Symbol::*;
        use TokenType::*;

        let input = "\
let foo = 14
// bar
/ not_a_comment
baz
";
        let mut lexer = Lexer::new(input);
        assert_eq!(Ok(Token::new(Let, 1, 1)), lexer.lex());
        assert_eq!(Ok(Token::new(Ident("foo".to_string()), 1, 5)), lexer.lex());
        assert_eq!(Ok(Token::new(Op(Assign), 1, 9)), lexer.lex());
        assert_eq!(Ok(Token::new(Int(14), 1, 11)), lexer.lex());
        assert_eq!(Ok(Token::new(Op(Div), 3, 1)), lexer.lex());
        assert_eq!(
            Ok(Token::new(Ident("not_a_comment".to_string()), 3, 3)),
            lexer.lex()
        );
        assert_eq!(Ok(Token::new(Ident("baz".to_string()), 4, 1)), lexer.lex());
        assert_eq!(Ok(Token::default()), lexer.lex());
    }
}
