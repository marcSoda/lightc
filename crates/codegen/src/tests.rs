use inkwell::context::Context;
use inkwell::passes::PassManager;

use hir::Hir;
use lexer::Lexer;
use parser::Parser;
use type_checker::TypeChecker;

use super::*;

macro_rules! run_insta {
    ($prefix:expr, $tests:expr) => {
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            for test in $tests {
                // Unoptimized code
                let tokens = Lexer::new(test[1]).scan().unwrap();
                let ast = Parser::new(&tokens).parse().unwrap();
                let mut symbol_cache = SymbolCache::new();
                let hir = Hir::new(&mut symbol_cache).walk(ast).unwrap();
                let thir = TypeChecker::new(&symbol_cache).walk(hir).unwrap();
                let context = Context::create();
                let builder = context.create_builder();
                let module = context.create_module("main_mod");
                let fpm = PassManager::create(&module);
                let codegen = Codegen::new(&context, &builder, &module, &fpm, &symbol_cache, 0, false, true);
                codegen.walk(thir).expect("codegen error");
                let res = module.print_to_string().to_string();

                // Optimized code
                let tokens = Lexer::new(test[1]).scan().unwrap();
                let ast = Parser::new(&tokens).parse().unwrap();
                let mut symbol_cache = SymbolCache::new();
                let hir = Hir::new(&mut symbol_cache).walk(ast).unwrap();
                let thir = TypeChecker::new(&symbol_cache).walk(hir).unwrap();
                let context = Context::create();
                let builder = context.create_builder();
                let module = context.create_module("main_mod");
                let fpm = PassManager::create(&module);
                let codegen = Codegen::new(&context, &builder, &module, &fpm, &symbol_cache, 1, false, true);
                codegen.walk(thir).expect("codegen error");
                let res_opt = module.print_to_string().to_string();

                insta::assert_yaml_snapshot!(format!("{}_{}", $prefix, test[0]), (test[1], res, res_opt));
            }
        })
    };
}

#[test]
fn test_block() {
    let tests = [[
        "basic",
        r#"
fn main() {
    {
        10 + 2
        1.0
        5
    }
}
"#,
    ]];
    run_insta!("block", tests);
}

#[test]
fn test_for() {
    let tests = [[
        "basic",
        r#"
extern putchar(x: int) -> int
fn main() {
    let x: int = 3
    for y: int = 0; y < x; 1 {
        putchar(46)
    }
    putchar(10)
}
"#,
    ]];
    run_insta!("for", tests);
}

#[test]
fn test_cond() {
    let tests = [
        [
            "cond_success",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 7 {
        7
    }
}
"#,
        ],
        [
            "cond_fail",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 5 {
        7
    }
}
"#,
        ],
        [
            "if_else_cond_success",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 7 {
        7
    } else {
        1
    }
}
"#,
        ],
        [
            "if_else_cond_fail",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 8 {
        7
    } else {
        1
    }
}
"#,
        ],
        [
            "if_else_if_else_cond_success",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 7 {
        7
    } else if plus_one(1) == 2 {
        2
    } else {
        1
    }
}
"#,
        ],
        [
            "if_else_if_else_cond_2_success",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 5 {
        7
    } else if plus_one(1) == 2 {
        2
    } else {
        1
    }
}
"#,
        ],
        [
            "if_else_if_else_conds_fail",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 5 {
        7
    } else if plus_one(1) == 3 {
        2
    } else {
        1
    }
}
"#,
        ],
        [
            "if_else_if_cond_success",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 7 {
        7
    } else if plus_one(1) == 1 {
        2
    }
}
"#,
        ],
        [
            "if_else_if_cond_2_success",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 5 {
        7
    } else if plus_one(1) == 2 {
        2
    }
}
"#,
        ],
        [
            "true_lit",
            r#"
fn main() {
    if true {
        7
    } else {
        2
    }
}
"#,
        ],
        [
            "false_lit",
            r#"
fn main() {
    if false {
        7
    } else {
        2
    }
}
"#,
        ],
    ];
    run_insta!("cond", tests);
}

// #[test]
// fn test_let() {
//     let tests = [
//         [
//             "basic",
//             r#"
//         fn main() {
//     let x: int = 1
// }
// "#,
//         ],
//         [
//             "float_default",
//             r#"
// fn main() {
//     let x: float
// }
// "#,
//         ],
//         [
//             "bool",
//             r#"
// fn main() {
//     let x: bool
// }
// "#,
//         ],
//     ];
//     run_insta!("let", tests);
// }

#[test]
fn test_scope() {
    let tests = [
        [
            "basic_shadowing",
            r#"
fn foo(a: int) -> int {
    let b: int = 1
    {
        let b: bool = false
    }
    b
}
fn main() { foo(3) }
"#,
        ],
        [
            "nested_shadowing",
            r#"
fn foo(a: int) -> int {
    let b: int = 1
    {
        let b: bool = false
        let a: float = {
            let b: float = 1.0
            b
        }
    }
    b
}
fn main() { foo(3) }
"#,
        ],
        [
            "for",
            r#"
fn foo() {
    let x: float = 1.0
    for x: int = 1; x < 10; 1 {
        x
    }
    x
}
fn main() { foo() }
"#,
        ],
        [
            "if",
            r#"
fn foo() {
    let x: float = 1.0
    if x < 2.0 {
        let y: int = 2
        x
    }
    x
}
fn main() { foo() }
"#,
        ],
        [
            "if_else",
            r#"
fn foo() {
    let x: float = 1.0
    if x < 2.0 {
        let y: int = 4 & 3 ^ 1
        y
    } else {
        let y: int = -2
        y
    }
}
fn main() { foo() }
"#,
        ],
    ];
    run_insta!("scope", tests);
}

#[test]
fn test_unop() {
    let tests = [[
        "basic",
        r#"
fn test(x: int) -> int {
    x + 1
}

fn main() -> int {
    test(-40)
}
"#,
    ]];
    run_insta!("unop", tests);
}

#[test]
fn test_array() {
    let tests = [[
        "basic",
        r#"
fn main() {
    let a: [int; 3] = [1, 2, 3]
    a[1] = 7
    a[1]
}
"#,
    ]];
    run_insta!("array", tests);
}
