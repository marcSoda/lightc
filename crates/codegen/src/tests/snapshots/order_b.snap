---
source: crates/codegen/src/tests.rs
assertion_line: 389
expression: "(test[1], res, res_opt)"
---
- "\nfn main() {\n    foo()\n}\nfn foo() {}\n"
- "; ModuleID = 'test'\nsource_filename = \"test\"\n\ndefine void @\"_foo~void\"() {\nentry:\n  ret void\n}\n\ndefine void @main() {\nentry:\n  call void @\"_foo~void\"()\n  ret void\n}\n"
- "; ModuleID = 'test'\nsource_filename = \"test\"\n\ndefine void @\"_foo~void\"() {\nentry:\n  ret void\n}\n\ndefine void @main() {\nentry:\n  call void @\"_foo~void\"()\n  ret void\n}\n"

